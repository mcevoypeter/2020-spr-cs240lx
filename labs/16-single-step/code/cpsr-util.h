#ifndef __CPSR_UTIL_H__
#define __CPSR_UTIL_H__

#include "rpi-interrupts.h"  // for cpsr_get, cpsr_set
#include "rpi-constants.h"  // for the modes.

static inline int mode_legal(unsigned mode) {
    return mode == USER_MODE
        || mode == IRQ_MODE
        || mode == SUPER_MODE
        || mode == ABORT_MODE
        || mode == SYS_MODE
    ;
}
static inline uint32_t mode_get(uint32_t cpsr) {
    unsigned mode = bits_get(cpsr, 0, 4);
    assert(mode_legal(mode));
    return mode;
}
static inline uint32_t mode_set(uint32_t cpsr, unsigned mode) {
    assert(mode_legal(mode));
    return bits_clr(cpsr, 0, 4) | mode;
}
static inline const char * mode_str(uint32_t cpsr) {
    switch(mode_get(cpsr)) {
    case USER_MODE: return "USER MODE";
    case FIQ_MODE:  return "FIQ MODE";
    case IRQ_MODE:  return "IRQ MODE";
    case SUPER_MODE: return "SUPER MODE";
    case ABORT_MODE: return "ABORT MODE";
    case UNDEF_MODE: return "UNDEF MODE";
    case SYS_MODE:   return "SYS MODE";
    default: panic("unknown mode: <%x>\n", cpsr);
    }
}
#endif
