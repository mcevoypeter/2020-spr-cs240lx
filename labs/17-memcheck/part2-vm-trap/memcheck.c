/*
 * engler: cs340lx
 *
 * start of a memchecker.  sort of lame in that you have to deal with a bunch of 
 * my starter code (sorry) --- the issue is that it might not be clear how to 
 * call the vm.
 *
 * implement:
 *  - static void dom_perm_set(unsigned dom, unsigned perm) {
 *  - void data_abort_vector(unsigned lr) {
 * (1) is trivial, but just to remind you about domain id's.
 * (2) involves computing the fault reason and then acting.
 *
 */
#include "memcheck.h"
#include "cp14-debug.h"
#include "libc/helper-macros.h"

enum { OneMB = 1024 * 1024 };

// don't use dom id = 0 --- too easy to miss errors.
enum { dom_id = 1, track_id = 2 };

/**********************************************************************
 * helper code to track the last fault (used for testing).
 */
static last_fault_t last_fault;
last_fault_t last_fault_get(void) {
    return last_fault;
}
void last_fault_set(uint32_t fault_pc, uint32_t fault_addr, uint32_t reason) {
    last_fault = (last_fault_t) { 
            .fault_addr = fault_addr, 
            .fault_pc = fault_pc,
            .fault_reason = reason
    };
}
void fault_expected(uint32_t pc, uint32_t addr, uint32_t reason) {
    last_fault_t l = last_fault_get();
    assert(l.fault_addr == addr);
    assert(l.fault_pc ==  pc);
    assert(l.fault_reason == reason);
    trace("saw expected fault at pc=%p, addr=%p, reason=%d\n", pc,addr,reason);
}

/****************************************************************
 * mechanical code for flipping permissions for the tracked memory
 * domain <track_id> on or off.
 *
   from in the armv6.h header:
    DOM_no_access = 0b00, // any access = fault.
    DOM_client = 0b01,      // client accesses check against permission bits in tlb
    DOM_reserved = 0b10,      // client accesses check against permission bits in tlb
    DOM_manager = 0b11,      // TLB access bits are ignored.
 */

static unsigned dom_perm_get(unsigned dom) {
    unsigned x = read_domain_access_ctrl();
    return bits_get(x, dom*2, (dom+1)*2);
}

// set the permission bits for domain id <dom> to <perm>
// leave the other domains the same.
static void dom_perm_set(unsigned dom, unsigned perm) {
    assert(dom < 16);
    assert(perm == DOM_client || perm == DOM_no_access);

    unimplemented();
}


/**************************************************************************
 * handle a trap.
 */

// hack to turn off resume / not resume for testing.
static unsigned mmu_resume_p = 0;
void memcheck_continue_after_fault(void) {
    mmu_resume_p = 1;
}

// simple data abort handle: handle the different reasons for the tests.
void data_abort_vector(unsigned lr) {
    if(was_debug_datafault())
        mem_panic("should not have this\n");

    // b4-43
    unsigned dfsr = dfsr_get();
    unsigned fault_addr = far_get();

    // compute the rason.
    unsigned reason = 0;
    unimplemented();

    last_fault_set(lr, fault_addr, reason);

    /*
         for SECTION_XLATE_FAULT:
            trace_clean_exit("section xlate fault: addr=%p, at pc=%p\n" 
         for DOMAIN_SECTION_FAULT:
            if(!mmu_resume_p)
                trace_clean_exit("Going to die!\n");
            else {
                trace("going to try to resume\n");
                memcheck_trap_disable();
                return;
            }
        for SECTION_PERM_FAULT:
            panic("section permission fault: %x", fault_addr);

        otherwise
        default: panic("unknown reason %b\n", reason);
     */ 
}

// shouldn't happen: need to fix libpi so we don't have to do this.
void interrupt_vector(unsigned lr) {
    mmu_disable();
    panic("impossible\n");
}


/*************************************************************************
 * helper code: you don't have to write this; it's based on cs140e.
 */

// XXX: our page table, gross.
static fld_t *pt = 0;

// need some parameters for this.
void memcheck_init(void) {
    // 1. init
    mmu_init();
    assert(!mmu_is_enabled());

    void *base = (void*)0x100000;

    pt = mmu_pt_alloc_at(base, 4096*4);
    assert(pt == base);

    // 2. setup mappings

    // map the first MB: shouldn't need more memory than this.
    mmu_map_section(pt, 0x0, 0x0, dom_id);
    // map the page table: for lab cksums must be at 0x100000.
    mmu_map_section(pt, 0x100000,  0x100000, dom_id);
    // map stack (grows down)
    mmu_map_section(pt, STACK_ADDR-OneMB, STACK_ADDR-OneMB, dom_id);

    // map the GPIO: make sure these are not cached and not writeback.
    // [how to check this in general?]
    mmu_map_section(pt, 0x20000000, 0x20000000, dom_id);
    mmu_map_section(pt, 0x20100000, 0x20100000, dom_id);
    mmu_map_section(pt, 0x20200000, 0x20200000, dom_id);

    // if we don't setup the interrupt stack = super bad infinite loop
    mmu_map_section(pt, INT_STACK_ADDR-OneMB, INT_STACK_ADDR-OneMB, dom_id);

    // 3. install fault handler to catch if we make mistake.
    mmu_install_handlers();

    // 4. start the context switch:

    // set up permissions for the different domains: we only use <dom_id>
    // and permissions r/w.
    write_domain_access_ctrl(0b01 << dom_id*2);

    // use the sequence on B2-25
    set_procid_ttbr0(0x140e, dom_id, pt);
}

// turn memchecking on: right now, just enable the mmu (which should 
// flush any needed caches).
void memcheck_on(void) {
    debug("memcheck: about to turn ON\n");
    // note: this routine has to flush I/D cache and TLB, BTB, prefetch buffer.
    mmu_enable();
    assert(mmu_is_enabled());
    debug("memcheck: ON\n");
}
// turn memcheck off: right now just diable mmu (which should flush
// any needed caches)
void memcheck_off(void) {
    // 6. disable.
    mmu_disable();
    assert(!mmu_is_enabled());
    debug("memcheck: OFF\n");
}

void memcheck_map(uint32_t base) {
    assert(is_aligned(base,OneMB));

    // XXX: need to handle when it's already mapped.
    mmu_map_section(pt, base, base + OneMB, dom_id);
}

void memcheck_track(uint32_t base) {
    assert(is_aligned(base,OneMB));

    // XXX: need to handle when it's already mapped.
    mmu_map_section(pt, base, base + OneMB, track_id);
}

void memcheck_no_access(uint32_t base) {
    base &= ~(OneMB - 1);
    assert(is_aligned(base,OneMB));
    mmu_mark_sec_no_access(pt, base, 1);
}

// is trapping enabled?
unsigned memcheck_trap_enabled(void) {
    return dom_perm_get(track_id) == DOM_no_access;
}

// disable traps on tracked memory: we still do permission based checks
// (you can disable this using DOM_manager).
void memcheck_trap_disable(void) {
    dom_perm_set(track_id, DOM_client);
    assert(!memcheck_trap_enabled());
}
void memcheck_trap_enable(void) {
    dom_perm_set(track_id, DOM_no_access);
    assert(memcheck_trap_enabled());
}
