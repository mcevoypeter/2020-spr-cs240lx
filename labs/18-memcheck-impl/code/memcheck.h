#ifndef __MEMCHECK_H__
#define __MEMCHECK_H__
#include "rpi.h"
#include "rpi-constants.h"
#include "rpi-interrupts.h"
#include "mmu.h"

// one-time initialization
void memcheck_init(void);

// start tracking.
void memcheck_on(void);
// stop
void memcheck_off(void);

// register the MB region [base,base+MB) so we start trapping on it.
void memcheck_track(uint32_t base);


// enable all memory trapping -- currently assumes using a single domain id.
void memcheck_trap_enable(void);

// disable all memory trapping -- currently assumes using a single domain id.
void memcheck_trap_disable(void);

// is trapping enabled?
unsigned memcheck_trap_enabled(void);


// XXX: hack to test that we can resume.
void memcheck_continue_after_fault(void);

int memcheck_fn(int (*fn)(void));

// for the moment, we just call these special allocator and free routines.
void *memcheck_alloc(unsigned n);
void memcheck_free(void *ptr);

void *sys_memcheck_alloc(unsigned n);
void sys_memcheck_free(void *ptr);

/**************************************************************
 * helper code: track information about the last fault.
 */

// a bit more general than memcheck, but we keep it here for simple.
typedef struct {
    uint32_t fault_addr,
             fault_pc,
             fault_reason;
} last_fault_t; 

// returns the last fault
last_fault_t last_fault_get(void);
// check that the last fault was at pc, using addr with <reason>
void fault_expected(uint32_t pc, uint32_t addr, uint32_t reason);

#include "cpsr-util.h"

#endif
