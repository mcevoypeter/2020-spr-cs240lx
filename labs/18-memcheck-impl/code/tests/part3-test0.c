// engler: check that allocation works with shadow memory: allocate, do a bunch
// of writes and read the result.
#include "memcheck.h"
#include "cpsr-util.h"

static void test_resume(int *addr, uint32_t val) {
    memcheck_alloc(4);

    trace("expect 'domain section fault' for addr %p, at pc %p\n", addr, put32);
    // should get a fault here.
    put32(addr, val);
    trace("checking that %p == %d\n", addr, val);

    assert(get32(addr) == val);
    trace("correct: %p == %d\n", addr, val);
}

unsigned expected = 45;

int notmain_client() {
    volatile int *x = memcheck_alloc(4);

    for(int i = 0; i < 10; i++)
        *x += i;
    printk("*x = %d\n", *x);
    assert(get32(x) == expected);
    return *x;
}

void notmain() {
    assert(!mmu_is_enabled());

    int x = memcheck_fn(notmain_client);
    assert(!mmu_is_enabled());

    assert(x == expected);


    trace_clean_exit("success!!\n");
}
