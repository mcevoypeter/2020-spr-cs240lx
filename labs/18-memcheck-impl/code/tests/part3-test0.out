find-ttyusb.c:find_ttyusb_first:106:FOUND: </dev/ttyUSB0>
opened tty port </dev/ttyUSB0>.
my-install: tty-usb=</dev/ttyUSB0> program=<./tests/part3-test0.bin> about to boot
******************sending 23296 bytes
waiting for a start
got a start
putting code
put code and got back
bootloader: Done.
listening on ttyusb=</dev/ttyUSB0>
memcheck.c:memcheck_on:140:memcheck: about to turn ON
memcheck.c:memcheck_on:144:memcheck: ON
TRACE:about to try running at user level
system call: 2
TRACE:domain section fault: 0x200000 at pc=0x803c
fault address = 0x200000
mismatch at pc 0x8040: disabling
load at pc=0x8040, addr=0x200000 checked out
legal load at pc=0x8040, addr=0x200000
TRACE:domain section fault: 0x200000 at pc=0x8044
fault address = 0x200000
mismatch at pc 0x8048: disabling
legal store at pc=0x8048, addr=0x200000
TRACE:domain section fault: 0x200000 at pc=0x803c
fault address = 0x200000
mismatch at pc 0x8040: disabling
load at pc=0x8040, addr=0x200000 checked out
legal load at pc=0x8040, addr=0x200000
TRACE:domain section fault: 0x200000 at pc=0x8044
fault address = 0x200000
mismatch at pc 0x8048: disabling
legal store at pc=0x8048, addr=0x200000
TRACE:domain section fault: 0x200000 at pc=0x803c
fault address = 0x200000
mismatch at pc 0x8040: disabling
load at pc=0x8040, addr=0x200000 checked out
legal load at pc=0x8040, addr=0x200000
TRACE:domain section fault: 0x200000 at pc=0x8044
fault address = 0x200000
mismatch at pc 0x8048: disabling
legal store at pc=0x8048, addr=0x200000
TRACE:domain section fault: 0x200000 at pc=0x803c
fault address = 0x200000
mismatch at pc 0x8040: disabling
load at pc=0x8040, addr=0x200000 checked out
legal load at pc=0x8040, addr=0x200000
TRACE:domain section fault: 0x200000 at pc=0x8044
fault address = 0x200000
mismatch at pc 0x8048: disabling
legal store at pc=0x8048, addr=0x200000
TRACE:domain section fault: 0x200000 at pc=0x803c
fault address = 0x200000
mismatch at pc 0x8040: disabling
load at pc=0x8040, addr=0x200000 checked out
legal load at pc=0x8040, addr=0x200000
TRACE:domain section fault: 0x200000 at pc=0x8044
fault address = 0x200000
mismatch at pc 0x8048: disabling
legal store at pc=0x8048, addr=0x200000
TRACE:domain section fault: 0x200000 at pc=0x803c
fault address = 0x200000
mismatch at pc 0x8040: disabling
load at pc=0x8040, addr=0x200000 checked out
legal load at pc=0x8040, addr=0x200000
TRACE:domain section fault: 0x200000 at pc=0x8044
fault address = 0x200000
mismatch at pc 0x8048: disabling
legal store at pc=0x8048, addr=0x200000
TRACE:domain section fault: 0x200000 at pc=0x803c
fault address = 0x200000
mismatch at pc 0x8040: disabling
load at pc=0x8040, addr=0x200000 checked out
legal load at pc=0x8040, addr=0x200000
TRACE:domain section fault: 0x200000 at pc=0x8044
fault address = 0x200000
mismatch at pc 0x8048: disabling
legal store at pc=0x8048, addr=0x200000
TRACE:domain section fault: 0x200000 at pc=0x803c
fault address = 0x200000
mismatch at pc 0x8040: disabling
load at pc=0x8040, addr=0x200000 checked out
legal load at pc=0x8040, addr=0x200000
TRACE:domain section fault: 0x200000 at pc=0x8044
fault address = 0x200000
mismatch at pc 0x8048: disabling
legal store at pc=0x8048, addr=0x200000
TRACE:domain section fault: 0x200000 at pc=0x803c
fault address = 0x200000
mismatch at pc 0x8040: disabling
load at pc=0x8040, addr=0x200000 checked out
legal load at pc=0x8040, addr=0x200000
TRACE:domain section fault: 0x200000 at pc=0x8044
fault address = 0x200000
mismatch at pc 0x8048: disabling
legal store at pc=0x8048, addr=0x200000
TRACE:domain section fault: 0x200000 at pc=0x803c
fault address = 0x200000
mismatch at pc 0x8040: disabling
load at pc=0x8040, addr=0x200000 checked out
legal load at pc=0x8040, addr=0x200000
TRACE:domain section fault: 0x200000 at pc=0x8044
fault address = 0x200000
mismatch at pc 0x8048: disabling
legal store at pc=0x8048, addr=0x200000
TRACE:domain section fault: 0x200000 at pc=0x8054
fault address = 0x200000
mismatch at pc 0x8058: disabling
load at pc=0x8058, addr=0x200000 checked out
legal load at pc=0x8058, addr=0x200000
*x = 45
TRACE:domain section fault: 0x200000 at pc=0xbe00
fault address = 0x200000
mismatch at pc 0xbe04: disabling
load at pc=0xbe04, addr=0x200000 checked out
legal load at pc=0xbe04, addr=0x200000
TRACE:domain section fault: 0x200000 at pc=0x8078
fault address = 0x200000
mismatch at pc 0x807c: disabling
load at pc=0x807c, addr=0x200000 checked out
legal load at pc=0x807c, addr=0x200000
system call: 1
done with user mode run!
memcheck.c:memcheck_off:152:memcheck: OFF
TRACE:success!!
DONE!!!

Saw done

bootloader: pi exited.  cleaning up
